### PR TITLE
XW-3303 | autoplay toggling

### PR DESCRIPTION
**Pull request by latata**
https://wikia-inc.atlassian.net/browse/XW-3303
@Wikia/x-wing

### CODE REVIEW

File: js/components/autoplaySwitch.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                    <span className={switchThumbClassName}></span>
+                </div>
+                <span className={onAutoplayClassName}></span>
+                <a className="oo-switch-container-selectable" onClick={this.handleOnOffSwitch}></a>
+            </div>
+        );
+    }
+});
+
+module.exports = AutoplaySwitch;
```
Comment (rybmat) [inline js/components/autoplaySwitch.js]: newline

File: js/components/configPanel.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+              {videoQualityItem}
+              <li className="oo-autoplay-element">{autoPlayText}<AutoplaySwitch autoPlay={this.props.controller.state.autoPlay} {...this.props} /></li>
+            </ul>
+          </div>
+        );
+    }
+});
+
+
+module.exports = ConfigPanel;
```
Comment (rybmat) [inline js/components/configPanel.js]: newline

File: scss/components/_autoplay.scss [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    background: $blue-light;
+  }
+  .oo-switch-body-off-autoplay {
+    background: $gray;
+  }
+}
+
+.oo-autoplay-element {
+  align-items: center;
+}
```
Comment (rybmat) [inline scss/components/_autoplay.scss]: newline

**Review by rybmat — APPROVED** [review]

File: js/components/configPanel.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    Utils = require('./utils'),
+    AutoplaySwitch = require('./autoplaySwitch');
+
+var ConfigPanel = React.createClass({
+    handleVideoQualityClick: function () {
+        this.props.toggleQualityAction();
+    },
+
+    render: function() {
+			var videoQualityText = Utils.getLocalizedString(this.props.language, CONSTANTS.SKIN_TEXT.VIDEO_QUALITY, this.props.localizableStrings);
```
Comment (rogatty) [inline js/components/configPanel.js]: We have spaces and tabs here, please unify.

File: js/components/configPanel.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/**
+ * Panel component for video quality selection
+ *
+ * @module VideoQualityPanel
```
Comment (rogatty) [inline js/components/configPanel.js]: This is a `ConfigPanel`, right?

File: js/components/configPanel.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    handleVideoQualityClick: function () {
+        this.props.toggleQualityAction();
+    },
+
+    render: function() {
+			var videoQualityText = Utils.getLocalizedString(this.props.language, CONSTANTS.SKIN_TEXT.VIDEO_QUALITY, this.props.localizableStrings);
+			var autoPlayText = Utils.getLocalizedString(this.props.language, CONSTANTS.SKIN_TEXT.AUTOPLAY_VIDEOS, this.props.localizableStrings);
+
+			var videoQualityItem;
+			if (this.props.controller.state.videoQualityOptions.availableBitrates) {
```
Comment (rogatty) [inline js/components/configPanel.js]: How about `var videoQualityItem = this.getVideoQualityItem()`? It's currently hard to read.

File: js/components/autoplaySwitch.js (hunk 1 above)
Comment (JamesHawking) [reply js/components/autoplaySwitch.js]: This is ooyala style.

File: js/components/controlBar.js [hunk 7]
```diff
@@ -112,13 +113,28 @@ var ControlBar = React.createClass({
     } else {
       this.toggleQualityPopover();
       this.closeCaptionPopover();
+      this.closeConfigPopover();
     }
   },
 
-  toggleQualityPopover: function() {
```
Comment (rogatty) [inline js/components/controlBar.js:118]: There are references to `toggleQualityPopover` in this file. Is that correct?

Comment (rogatty) [reply js/components/controlBar.js:118]: What's the reason to rename this method? It calls `toggleVideoQualityPopOver` anyway.

File: js/components/controlBar.js [hunk 8]
```diff
@@ -269,8 +285,9 @@ var ControlBar = React.createClass({
       "oo-control-bar-item oo-live oo-live-indicator": true,
       "oo-live-nonclickable": isLiveNow
     });
-
-    var videoQualityPopover = this.props.controller.state.videoQualityOptions.showVideoQualityPopover ? <Popover><VideoQualityPanel{...this.props} togglePopoverAction={this.toggleQualityPopover} popover={true}/></Popover> : null;
+    var showVideoQualityPanel = !this.props.skinConfig.controlBar.autoplayToggle || this.props.controller.state.configPanelOptions.showVideoQualityPanel;
+    var configPanelContent = showVideoQualityPanel ? <VideoQualityPanel{...this.props} togglePopoverAction={this.toggleConfigPopover} toggleVideoQualityPanel={this.toggleQualityPanel} popover={true}/> :  <ConfigPanel {...this.props} toggleQualityAction={this.toggleQualityPanel} />;
+    var configPanelPopover = this.props.controller.state.configPanelOptions.showConfigPanelPopover  ? <Popover popoverClassName="oo-popover oo-popover-pull-autoplay">{configPanelContent}</Popover> : null;
```
Comment (rogatty) [inline js/components/controlBar.js]: Does the name `oo-popover-pull-autoplay` make sense? `oo-popover-pull-right` means it's pulled to the right, right?

File: js/components/controlBar.js (hunk 7 above)
Comment (JamesHawking) [reply js/components/controlBar.js:118]: This ooyala way of toggling, wanted to be consistent here.

File: js/controller.js [hunk 9]
```diff
@@ -19,123 +20,142 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
     return null;
   }
 
-  if (OO.publicApi && OO.publicApi.VERSION) {
+	if (OO.publicApi && OO.publicApi.VERSION) {
```
Comment (rogatty) [inline js/controller.js]: This file needs some reformatting too.

File: js/controller.js [hunk 10]
```diff
@@ -19,123 +20,142 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
...
-      "autoPauseDisabled": false
-    };
-
-    this.init();
+  OO.EVENTS.WIKIA = {
+    AUTOPLAY_TOGGLED: 'wikia.autoplayToggled'
+  };
+  OO.exposeStaticApi('EVENTS', OO.EVENTS);
+
+	var autoplayCookieName = 'html5-skin.autoplay',
```
Comment (rogatty) [inline js/controller.js]: Didn't we want it to be configurable? Or is there currently no need for that?

File: js/controller.js [hunk 11]
```diff
@@ -1358,6 +1392,14 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
       this.mb.publish(OO.EVENTS.SAVE_PLAYER_SETTINGS, this.state.persistentSettings);
     },
 
+    toggleAutoPlayEnabled: function() {
+      this.state.autoPlay.enabled = !this.state.autoPlay.enabled;
+      this.state.persistentSettings.autoPlay['enabled'] = !!this.state.autoPlay.enabled;
+      this.renderSkin();
+      Cookies.set(autoplayCookieName, this.state.autoPlay.enabled ? 1 : 0, { expires: 14 });
+      this.mb.publish(OO.EVENTS.WIKIA.AUTOPLAY_TOGGLED, this.state.autoPlay.enabled);
```
Comment (rogatty) [inline js/controller.js]: Should we also publish `OO.EVENTS.SAVE_PLAYER_SETTINGS`?

File: js/controller.js [hunk 12]
```diff
@@ -1358,6 +1392,14 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
       this.mb.publish(OO.EVENTS.SAVE_PLAYER_SETTINGS, this.state.persistentSettings);
     },
 
+    toggleAutoPlayEnabled: function() {
+      this.state.autoPlay.enabled = !this.state.autoPlay.enabled;
+      this.state.persistentSettings.autoPlay['enabled'] = !!this.state.autoPlay.enabled;
+      this.renderSkin();
+      Cookies.set(autoplayCookieName, this.state.autoPlay.enabled ? 1 : 0, { expires: 14 });
```
Comment (rogatty) [inline js/controller.js]: I think 14 should be configurable.

File: scss/components/_popover.scss [hunk 13]
```diff
@@ -14,6 +14,12 @@
     @include vendor-prefixes(transform, none);
   }
 
+  &.oo-popover-pull-autoplay {
```
Comment (rogatty) [inline scss/components/_popover.scss]: I believe this is obsolete.

File: js/controller.js [hunk 14]
```diff
@@ -1358,6 +1398,14 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
       this.mb.publish(OO.EVENTS.SAVE_PLAYER_SETTINGS, this.state.persistentSettings);
     },
 
+    toggleAutoPlayEnabled: function() {
+      this.state.autoPlay.enabled = !this.state.autoPlay.enabled;
+      this.state.persistentSettings.autoPlay['enabled'] = !!this.state.autoPlay.enabled;
+      this.renderSkin();
+      Cookies.set(autoplayCookieName, this.state.autoPlay.enabled ? 1 : 0, { expires: autoplayCookieExpireDays });
+      this.mb.publish(OO.EVENTS.WIKIA.AUTOPLAY_TOGGLED, this.state.autoPlay.enabled);
```
Comment (rogatty) [inline js/controller.js]: Should we also publish `OO.EVENTS.SAVE_PLAYER_SETTINGS`?

**Review by rogatty — APPROVED** [review]
Two minor things but generally LGTM

File: js/controller.js (hunk 14 above)
Comment (latata) [reply js/controller.js]: we shouldn't. It only saves closed captions settings.

**Pull request merged**